crates/paralegal-policy/src/flows_to.rs

```diff
@@ -47,54 +62,100 @@ impl CtrlFlowsTo {
+                // initialize with flows from DataSource to the DataSink's CallSite
+                if let DataSink::Argument { function, .. } = sink {
+                    data_flows_to.insert(src, CallSiteOrDataSink::CallSite(function.clone()));
+                }
             }
         }
 
-        // Compute the transitive closure to a fixpoint.
-        loop {
-            let mut changed = false;
-
-            for (src_idx, _src) in sources.as_vec().iter_enumerated() {
-                for sink_idx in flows_to.row_set(&src_idx).indices().collect::<Vec<_>>() {
-                    for trans_src_idx in sink_to_source.row_set(&sink_idx).indices() {
-                        for trans_sink_idx in flows_to
-                            .row_set(&trans_src_idx)
-                            .indices()
-                            .collect::<Vec<_>>()
-                        {
-                            changed |= flows_to.insert(src_idx, trans_sink_idx);
-                        }
+        iterate(&sources, &mut data_flows_to, &sink_to_source);
+
+        // Initialize the `flows_to` relation with the data provided by `Ctrl::data_flow`.
+        let mut flows_to = data_flows_to.clone();
+
+        // Connect each callsite to its function-argument sinks
+        let mut cs_to_sink = IndexMatrix::new(&sinks);
+        for (callsite_idx, callsite) in sinks.as_vec().iter_enumerated() {
+            for (sink_idx, sink) in sinks.as_vec().iter_enumerated() {
+                if let (
+                    CallSiteOrDataSink::DataSink(DataSink::Argument { function: f1, .. }),
+                    CallSiteOrDataSink::CallSite(f2),
+                ) = (sink, callsite)
+                {
+                    if f1 == f2 {
+                        cs_to_sink.insert(callsite_idx, sink_idx);
                     }
                 }
             }
-
-            if !changed {
-                break;
+        }
+        // Initialize the `flows_to` relation with the data provided by `Ctrl::ctrl_flow`.
+        for (src, callsites) in &ctrl.ctrl_flow.0 {
+            let src = src.to_index(&sources);
+            for cs in callsites {
+                let new_call_site = CallSiteOrDataSink::CallSite(cs.clone());
+                flows_to.insert(src, &new_call_site);
+                // initialize with flows from the DataSource to all of the CallSite's DataSinks
+                for sink in cs_to_sink.row_set(&sinks.index(&new_call_site)).iter() {
+                    flows_to.insert(src, sink);
+                }
             }
         }
 
+        iterate(&sources, &mut flows_to, &sink_to_source);
+
         CtrlFlowsTo {
             sources,
             sinks,
+            data_flows_to,
             flows_to,
         }
     }
 }
 
 impl fmt::Debug for CtrlFlowsTo {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        self.flows_to.fmt(f)
+        self.data_flows_to.fmt(f)
```

Comment:
I'm not happy with this implementation. I'd like to see what happens if we remove this `impl` altogether. 
But if it is used, then we should make sure this prints the data and the mixed flow. You can use the [`debug_struct`](https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.debug_struct) interface to create a debug dump of both fields.

Reply:
Used to impl Debug for Context. Done!